src/data/finvec.lean

```diff
@@ -0,0 +1,193 @@
+append_equiv.symm (x, y)
+
+localized "infixr ` ++ ` := finvec.append" in finvec
+
+lemma append_equiv_app {z : finvec (n + m) α} : append_equiv z = (z.left, z.right) := rfl
+
+lemma append_equiv_symm_app {x : finvec n α} {y : finvec m α} :
+  append_equiv.symm (x, y) = x ++ y :=
+rfl
+
+@[simp] lemma left_append_right (x : finvec (n + m) α) : x.left ++ x.right = x :=
+append_equiv.symm_apply_apply x
+
+/-- Induction principle for `finvec (n + m) α`,
+imagined as a structure containing its left and right parts. -/
+@[elab_as_eliminator]
+protected lemma prod_induction {C : finvec (n + m) α → Prop}
+  (h : Π (x : finvec n α) (y : finvec m α), C (x ++ y)) (z : finvec (n + m) α) : C z :=
+left_append_right z ▸ h z.left z.right
+
+@[simp] lemma left_append {x : finvec n α} {y : finvec m α} : (x ++ y).left = x :=
+congr_arg prod.fst (append_equiv.apply_symm_apply (x, y) : _)
+
+@[simp] lemma right_append {x : finvec n α} {y : finvec m α} : (x ++ y).right = y :=
+congr_arg prod.snd (append_equiv.apply_symm_apply (x, y) : _)
+
+lemma prod_ext {z z' : finvec (n + m) α} :
+  z = z' ↔ z.left = z'.left ∧ z.right = z'.right :=
+append_equiv.apply_eq_iff_eq.symm.trans prod.mk.inj_iff
+
+lemma append.inj_iff {x x' : finvec n α} {y y' : finvec m α} :
+  x ++ y = x' ++ y' ↔ x = x' ∧ y = y' :=
+by simp only [append, equiv.apply_eq_iff_eq, prod.mk.inj_iff]
+
+lemma left_zero {x : finvec (n + 0) α} : x.left = x :=
+by { ext ⟨_, _⟩, refl }
+
+lemma right_zero {x : finvec (0 + n) α} : x.right = x.cast (zero_add n) :=
+funext $ λ _, congr_arg x (by rw fin.nat_add_zero)
+
+@[simp] lemma append_zero {x : finvec n α} {y : finvec 0 α} : x ++ y = x :=
+begin
+  rw [prod_ext, left_append, left_zero, right_append],
+  exact ⟨rfl, subsingleton.elim _ _⟩
+end
+
+end prod
+
+section snoc
+
+/-! ### Appending a single element (snoc)
+
+This section is parallel to the `prod` section but based on the isomorphism αⁿ⁺¹ ≅ αⁿ × α.
+We name the projections `init : finvec (n+1) α → finvec n α`
+(which equals `left` specialized to `m = 1`) and `last : finvec (n+1) α → α`,
+with constructor `snoc : finvec n α → α → finvec (n+1) α` and
+corresponding induction principle `snoc_induction`. -/
+
+variables {n : ℕ}
+
+/-- All but the last coordinate of a `finvec`. -/
+protected def init : finvec (n + 1) α → finvec n α :=
+λ x j, x (fin.cast_succ j)
+
+lemma left_eq_init {x : finvec (n + 1) α} : x.left = x.init := rfl
+
+/-- The last coordinate of a `finvec`. -/
+protected def last : finvec (n + 1) α → α :=
+λ x, x (fin.last n)
+
+/-- The canonical equivalence `αⁿ⁺¹ ≅ αⁿ × α`. -/
+def snoc_equiv : finvec (n + 1) α ≃ finvec n α × α :=
+calc  (fin (n + 1) → α)
+    ≃ (fin n → α) × (fin 1 → α)  : append_equiv
+... ≃ (fin n → α) × α            : equiv.prod_congr (equiv.refl _) (equiv.fun_unique (fin 1) α)
+
+/-- Form a `finvec` by appending a single coordinate to an existing `finvec`. -/
+def snoc (x : finvec n α) (a : α) : finvec (n + 1) α :=
```

Comment:
I'm not happy about all of the duplication.  We already have a lot of these definitions for `fin n → α`, this one is `fin.snoc`.

If we introduce a `finvec` type, then the existing usage of `fin n → α` should be migrated.  Otherwise we just have yet another container type.

Rob's proposal of making `finvec` reducible sounds good to me.  Maybe we should also have a `dfinvec` for the dependently-typed version.  Then we could move all the `fin.cons`, etc., definitions to `dfinvec` and `finvec` without breaking too much.

Reply:
> I'm not happy about all of the duplication. We already have a lot of these definitions for `fin n → α`, this one is `fin.snoc`.

`fin.snoc` is a poor substitute for this `finvec.snoc` because of its more complicated type. For example in https://github.com/rwbarton/lean-omin/commit/b9a33a7899f5b990d1e6be25a8c145bf66853dea I replaced the theorem statement
```lean
lemma mem_image_fin_init {n : ℕ} {s : set (finvec (n + 1) R)} {y : finvec n R} :
  y ∈ s.image fin.init ↔ ∃ z : R, (fin.snoc y z : fin (n + 1) → R) ∈ s := [...]
```
by
```lean
lemma mem_image_init {s : set (finvec (n + 1) α)} {y : finvec n α} :
  y ∈ finvec.init '' s ↔ ∃ z : α, finvec.snoc y z ∈ s := [...]
```
IIRC the changes on both sides of the `↔` were impossible with `fin.snoc`.

> If we introduce a `finvec` type, then the existing usage of `fin n → α` should be migrated. Otherwise we just have yet another container type.
> 
> Rob's proposal of making `finvec` reducible sounds good to me. Maybe we should also have a `dfinvec` for the dependently-typed version. Then we could move all the `fin.cons`, etc., definitions to `dfinvec` and `finvec` without breaking too much.

I like the idea of making `finvec` reducible too, and it might have made my own transition to `finvec` in lean-omin a bit smoother.

Adding `dfinvec` would be a good next step, and would hopefully replace the existing uses of things like `fin.snoc`. One tricky thing is that the logical thing to index `dfinvec` on is a `finvec` of types, in which case the non-dependent version must exist first anyways.